docs/source/python/extending_types.rst

```diff
@@ -44,3 +44,162 @@ define the ``__arrow_array__`` method to return an Arrow array::
+  recognized by other (non-Python) Arrow implementations such as PySpark.
+
+For example, we could define a custom UUID type for 128-bit numbers which can
+be represented as ``FixedSizeBinary`` type with 16 bytes.
+Using the first approach, we create a ``UuidType`` subclass, and implement the
+``__reduce__`` method to ensure the class can be properly pickled::
+
+    class UuidType(pa.PyExtensionType):
+
+        def __init__(self):
+            pa.PyExtensionType.__init__(self, pa.binary(16))
+
+        def __reduce__(self):
+            return UuidType, ()
+
+This can now be used to create arrays and tables holding the extension type::
+
+    >>> uuid_type = UuidType()
+    >>> uuid_type.extension_name
+    'arrow.py_extension_type'
+    >>> uuid_type.storage_type
+    FixedSizeBinaryType(fixed_size_binary[16])
+
+    >>> import uuid
+    >>> storage_array = pa.array([uuid.uuid4().bytes for _ in range(4)], pa.binary(16))
+    >>> arr = pa.ExtensionArray.from_storage(uuid_type, storage_array)
+    >>> arr
+    <pyarrow.lib.ExtensionArray object at 0x7f75c2f300a0>
+    [
+      A6861959108644B797664AEEE686B682,
+      718747F48E5F4058A7261E2B6B228BE8,
+      7FE201227D624D96A5CD8639DEF2A68B,
+      C6CA8C7F95744BFD9462A40B3F57A86C
+    ]
+
+This array can be included in RecordBatches, sent over IPC and received in
+another Python process. The custom UUID type will be preserved there, as long
+as the definition of the class is available (the type can be unpickled).
+
+We can define the same type using the other option::
+
+    class UuidType(pa.ExtensionType):
+
+        def __init__(self):
+            pa.ExtensionType.__init__(self, pa.binary(16), "my_package.uuid")
+
+        def __arrow_ext_serialize__(self):
+            # since we don't have a parametrized type, we don't need extra
+            # metadata to be deserialized
+            return b''
+
+        @classmethod
+        def __arrow_ext_deserialize__(self, storage_type, serialized):
+            # return an instance of this subclass given the serialized
+            # metadata.
+            return UuidType()
+
+This is a slightly longer implementation (you need to implement the special
+methods ``__arrow_ext_serialize__`` and ``__arrow_ext_deserialize__``), and the
+extension type needs to be registered to be received through IPC, but it has
+now a unique name::
+
+    >>> uuid_type = UuidType()
+    >>> uuid_type.extension_name
+    'my_package.uuid'
+
+    >>> pa.lib.register_extension_type(uuid_type)
+
+The receiving application doesn't need to be Python but can still recognize
+the extension type as a "uuid" type, if it has implemented its own extension
+type to receive it.
+
+Parametrized extension type
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The above example used a fixed storage type with no further metadata. But
+more flexible, parametrized extension types are also possible.
+
+The example given here implements an extension type for the pandas "period"
```

Comment:
Add a link to Pandas or its period datatype? Also, should "Pandas" be capitalized?